File: tools/sweep.py
#!/usr/bin/env python3
"""Dumps what this library answers for every ZIP code there is.

All 1,000 three-digit prefixes and all 100,000 five-digit codes, one per line.
Its twin tools/sweep.go prints the same thing from the Go side, and the two
files are compared -- which is the only way to know the two lookups agree on
inputs nobody thought to write a case for.

    python3 tools/sweep.py > /tmp/py.sweep
"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import ziptz  # noqa: E402  (after the path, on purpose)


def say(token):
    """One answer, error text included: what it refuses and why is as much a
    part of the port as what it resolves.

    prefix_zone and exact_zone are swept alongside zone and generic because
    they are public and because they are the two that answer "" rather than
    raising -- so a disagreement between the ports shows up as an empty column
    here and nowhere else. Location and Abbrev are left out on purpose: they
    read the system tz database, which makes them a test of the machine as much
    as of the tables, and Abbrev over 101,000 tokens costs half a minute of
    LoadLocation on the Go side for an answer Zone has already settled.
    """
    tail = f"{ziptz.prefix_zone(token[:3])}\t{ziptz.exact_zone(token)}"
    try:
        return f"{token}\t{ziptz.zone(token)}\t{ziptz.generic(token)}\t{tail}\n"
    except ziptz.ZipError as exc:
        return f"{token}\t!{exc}\t{tail}\n"


def main():
    out = sys.stdout
    out.writelines(say(f"{n:03d}") for n in range(1000))
    out.writelines(say(f"{n:05d}") for n in range(100000))


if __name__ == "__main__":
    main()

File: tools/sweep.go
//go:build ignore

// Dumps what this library answers for every ZIP code there is: all 1,000
// three-digit prefixes and all 100,000 five-digit codes, one per line. Its twin
// tools/sweep.py prints the same thing from the Python side, and the two files
// are compared -- which is the only way to know the two lookups agree on inputs
// nobody thought to write a case for.
//
//	go run tools/sweep.go > /tmp/go.sweep
package main

import (
	"bufio"
	"fmt"
	"os"

	"github.com/choey/ziptz"
)

func main() {
	out := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer out.Flush()
	for n := 0; n < 1000; n++ {
		say(out, fmt.Sprintf("%03d", n))
	}
	for n := 0; n < 100000; n++ {
		say(out, fmt.Sprintf("%05d", n))
	}
}

// say prints one answer, error text included: what it refuses and why is as
// much a part of the port as what it resolves.
// PrefixZone and ExactZone are swept alongside Zone and Generic because they
// are public and because they are the two that answer "" rather than an error
// -- so a disagreement between the ports shows up as an empty column here and
// nowhere else. Location and Abbrev are left out on purpose: they read the
// system tz database, which makes them a test of the machine as much as of the
// tables, and Abbrev over 101,000 tokens costs half a minute of LoadLocation
// for an answer Zone has already settled.
func say(out *bufio.Writer, token string) {
	tail := ziptz.PrefixZone(token[:3]) + "\t" + ziptz.ExactZone(token)
	name, err := ziptz.Zone(token)
	if err != nil {
		fmt.Fprintf(out, "%s\t!%s\t%s\n", token, err, tail)
		return
	}
	generic, err := ziptz.Generic(token)
	if err != nil {
		fmt.Fprintf(out, "%s\t!%s\t%s\n", token, err, tail)
		return
	}
	fmt.Fprintf(out, "%s\t%s\t%s\t%s\n", token, name, generic, tail)
}
